src/mumble/ALSAAudio.cpp

```diff
@@ -379,21 +378,27 @@ void ALSAAudioInput::run() {
 		snd_pcm_status_dump(status, log);
 		snd_pcm_status_free(status);
 #endif
-		readblapp = snd_pcm_readi(capture_handle, inbuff.data(), static_cast< snd_pcm_uframes_t >(wantPeriod));
-		if (readblapp == -ESTRPIPE) {
-			qWarning("ALSAAudioInput: PCM suspended, trying to resume");
-			while (bRunning && snd_pcm_resume(capture_handle) == -EAGAIN)
-				msleep(1000);
-			if ((err = snd_pcm_prepare(capture_handle)) < 0)
-				qWarning("ALSAAudioInput: %s: %s", snd_strerror(static_cast< int >(readblapp)), snd_strerror(err));
-		} else if (readblapp == -EPIPE) {
-			err = snd_pcm_prepare(capture_handle);
-			qWarning("ALSAAudioInput: %s: %s", snd_strerror(static_cast< int >(readblapp)), snd_strerror(err));
-		} else if (readblapp < 0) {
-			err = snd_pcm_prepare(capture_handle);
-			qWarning("ALSAAudioInput: %s: %s", snd_strerror(static_cast< int >(readblapp)), snd_strerror(err));
-		} else if (wantPeriod == static_cast< unsigned int >(readblapp)) {
-			addMic(inbuff.data(), static_cast< unsigned int >(readblapp));
+		const snd_pcm_sframes_t ret = snd_pcm_readi(capture_handle, inbuff.data(), wantPeriod);
+		if (ret >= 0) {
+			if (static_cast< snd_pcm_uframes_t >(ret) == wantPeriod) {
+				addMic(inbuff.data(), static_cast< unsigned int >(ret));
+			}
+		} else {
+			err = static_cast< decltype(err) >(ret);
+			switch (err) {
+				case -EINTR:
+				case -EPIPE:
+				case -ESTRPIPE:
+					qWarning("ALSAAudioInput: %s, attempting to recover the stream...", snd_strerror(err));
+					while (bRunning && snd_pcm_recover(capture_handle, err, 1) == -EAGAIN) {
+						msleep(1000);
```

Comment:
Maybe [sleep_for](https://en.cppreference.com/w/cpp/thread/sleep_for) should be preferred?

Reply:
Definitely, it's part of the C++ standard.

Reply:
Nevermind, I just realized the `msleep()` we are using is Qt's:

https://doc.qt.io/qt-6/qthread.html#msleep
![](https://github.com/user-attachments/assets/0892fbae-0c69-4471-86d1-b5c7bfc953a6)

Since we are using a `QThread` I would keep using its function for consistency.

Please note that there is also a `sleep()` function that is simply a wrapper treating the argument value as seconds:

![](https://github.com/user-attachments/assets/b36bba01-69c0-4b80-8e1c-19ec71527d57)